CVLByExample/structs/BankAccounts/certora/specs/BankAccounts.spec

```diff
@@ -0,0 +1,105 @@
+/***
+ * # Structs Example
+ *
+ * This is an example specification for using structs.
+ */
+
+methods {
+    function balanceOf(address)        external returns(uint) envfree;
+    function balanceOfAccount(address a, uint account) external returns(uint) envfree;
+    function totalSupply()             external returns(uint) envfree;
+    function getCustomer(address a) external returns(BankAccounts.Customer) envfree;
+}
+
+//// ## Part 1: Basic rules ////////////////////////////////////////////////////
+
+rule sanity {
+  env e;
+  calldataarg arg;
+  method f;
+  f(e, arg);
+  satisfy true;
+}
+
+// Transfer must move `amount` tokens from the caller's account to `recipient`
+rule transferSpec() {
+    address sender; address recip; uint amount; uint senderAccount; uint toAccount;
+
+    env e;
+    require e.msg.sender == sender;
+
+    mathint balance_sender_before = balanceOf(sender);
+    mathint balance_recip_before = balanceOf(recip);
+
+    transfer(e, recip, amount, senderAccount, toAccount);
+
+    mathint balance_sender_after = balanceOf(sender);
+    mathint balance_recip_after = balanceOf(recip);
+
+    require sender != recip;
+
+    assert balance_sender_after == balance_sender_before - amount,
+        "transfer must decrease sender's balance by amount";
+
+    assert balance_recip_after == balance_recip_before + amount,
+        "transfer must increase recipient's balance by amount";
+}
+
+
+/// Transfer must revert if the sender's balance is too small
+rule transferReverts {
+    env e; address recip; uint amount; uint fromAccount; uint toAccount;
+
+    require balanceOfAccount(e.msg.sender, fromAccount) < amount;
+
+    transfer@withrevert(e, recip, amount, fromAccount, toAccount);
+
+    assert lastReverted,
+        "transfer(recip,amount) must revert if sender's balance is less than `amount`";
+}
+
+
+/// Transfer must not revert unless
+///  the sender doesn't have enough funds,
+///  or the message value is nonzero,
+///  or the recipient's balance would overflow,
+///  or the message sender is 0,
+///  or the recipient is 0
+///
+/// @title Transfer doesn't revert
+rule transferDoesntRevert(BankAccounts.Customer c) {
+    env e; address recipient; uint amount; uint fromAccount; uint toAccount;
+
+    require fromAccount < c.accounts.length;
+
+    // BankAccounts.Customer c2 = getCustomer(recipient);
+    // require toAccount < c2.accounts.length;
+    require c.accounts[fromAccount].balance > amount;
```

Comment:
nice, this cannot be done with an assignment right? maybe have it with a comment that this is not supported to avoid overriding